MAD.Extensions.EFCore/UpsertExtensions.cs

```diff
@@ -11,32 +11,51 @@ namespace MAD.Extensions.EFCore
 {
     public static class UpsertExtensions
     {
-        public static void Upsert(this DbContext dbContext, object entity, Action<object> transformations = null)
+        public static void Upsert(this DbContext dbContext, object entity)
         {
             dbContext.ChangeTracker.TrackGraph(entity, g =>
             {
-                var entity = g.Entry.Entity;
                 var entityType = g.Entry.OriginalValues.EntityType;
 
                 if (entityType is null)
                     return;
 
-                transformations?.Invoke(entity);
+                if (g.InboundNavigation is null == false && g.InboundNavigation.IsCollection())
+                {
+                    var principalKeyProperties = g.InboundNavigation.ForeignKey.PrincipalKey.Properties;
+                    var dependantKeyProperties = g.InboundNavigation.ForeignKey.Properties;
+
+                    var pkValues = new List<object>();
+
+                    foreach (var pk in principalKeyProperties)
+                    {
+                        var value = dbContext.Entry(entity).Property(pk.Name).CurrentValue;
+                        pkValues.Add(value);
+                    }
+                    
+                    for (int i = 0; i < pkValues.Count; i++)
+                    {
+                        var dk = dependantKeyProperties[i];
+                        var pkVal = pkValues[i];
+
+                        dbContext.Entry(g.Entry.Entity).Property(dk.Name).CurrentValue = pkVal;
+                    }
+                }               
 
                 var primaryKey = entityType.FindPrimaryKey();
                 var keys = primaryKey.Properties.ToDictionary(
                     keySelector: y => y.Name,
                     elementSelector: x =>
                     {
-                        object result;
+                        object result = null;
```

Comment:
Don't assign null as a default value, leave as object result;

Reply:
Removed the default value assignment